pkg/auth/providers/openldap/openldap_client.go

```diff
@@ -0,0 +1,447 @@
+package openldap
+
+import (
+	"crypto/x509"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/norman/httperror"
+	"github.com/rancher/norman/types/slice"
+	"github.com/rancher/rancher/pkg/auth/providers/common/ldap"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/apis/management.cattle.io/v3public"
+	"github.com/sirupsen/logrus"
+	ldapv2 "gopkg.in/ldap.v2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (p *openldapProvider) loginUser(credential *v3public.BasicLogin, config *v3.OpenLDAPConfig, caPool *x509.CertPool) (v3.Principal, []v3.Principal, map[string]string, error) {
+	logrus.Debug("Now generating Ldap token")
+
+	username := credential.Username
+	password := credential.Password
+
+	if password == "" {
+		return v3.Principal{}, nil, nil, httperror.NewAPIError(httperror.MissingRequired, "password not provided")
+	}
+
+	lConn, err := ldap.NewLDAPConnForOpenLDAP(config, caPool)
+	if err != nil {
+		return v3.Principal{}, nil, nil, err
+	}
+	defer lConn.Close()
+
+	if !config.Enabled { // TODO testing for enabled here might not be correct. Might be better to pass in an explicit testSvcAccount bool
+		logrus.Debug("Bind service account username password")
+		if config.ServiceAccountPassword == "" {
+			return v3.Principal{}, nil, nil, httperror.NewAPIError(httperror.MissingRequired, "service account password not provided")
+		}
+		sausername := ldap.GetUserExternalID(config.ServiceAccountUsername, "")
+		err = lConn.Bind(sausername, config.ServiceAccountPassword)
+		if err != nil {
+			if ldapv2.IsErrorWithCode(err, ldapv2.LDAPResultInvalidCredentials) {
+				return v3.Principal{}, nil, nil, httperror.WrapAPIError(err, httperror.Unauthorized, "authentication failed")
+			}
+			return v3.Principal{}, nil, nil, httperror.WrapAPIError(err, httperror.ServerError, "server error while authenticating")
+		}
```

Comment:
Can this be moved to a common util method say authenticateServiceUser() and called from both AD and openLDAP?